### PR TITLE
implement Style/DisableCopsWithinSourceCodeDirective cop

### PR DESCRIPTION
**Pull request by egze**
This cop forbids enabling or disabling cops within source code with comments like:
```ruby
# rubocop:disable Style/EmptyBlockParameter
```
The motivation behind it is to make sure developers fix the issues reported by rubocop and not quickly add comments to ignore them.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by bbatsov** [comment]
I like the spirit of the change, but I'm not sure if this should be a cop or some global config in RuboCop. @jonas054 What do you think on the subject?

File: lib/rubocop/cop/style/disable_cops_within_source_code_directive.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      #   def fixed_method_name_and_no_rubocop_comments
+      #   end
+      #
+      class DisableCopsWithinSourceCodeDirective < Cop
+        # rubocop:enable Lint/UnneededCopDisableDirective
+        MSG = 'Comment to disable/enable Rubocop.'
+
+        def investigate(processed_source)
+          processed_source.comments.each do |comment|
+            next unless comment.text.start_with?('# rubocop:')
```
Comment (jonas054) [inline lib/rubocop/cop/style/disable_cops_within_source_code_directive.rb]: There's a constant, `CommentConfig::COMMENT_DIRECTIVE_REGEXP` that you could use instead to catch all variations on whitespace that we allow in these comments.

**Comment by jonas054** [comment]
I love the idea! In fact, I argued rather stubbornly for implementing this kind of functionality in a cop back in #900, but I was overruled and and the `--format disabled` CLI option was added instead. This option doesn't currently work (I just discovered), and hasn't worked for a long time. That means that people aren't using it.

So, I like the cop and the changes look good (had some minor comments), and I would like to see the `DisabledLinesFormatter` gone as well.

File: lib/rubocop/cop/style/disable_cops_within_source_code_directive.rb (hunk 1 above)
Comment (egze) [reply lib/rubocop/cop/style/disable_cops_within_source_code_directive.rb]: Done

File: lib/rubocop/cop/style/disable_cops_within_source_code_directive.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      #   end
+      #   # rubocop:enable Metrics/AbcSize
+      #
+      #   # good
+      #   def fixed_method_name_and_no_rubocop_comments
+      #   end
+      #
+      class DisableCopsWithinSourceCodeDirective < Cop
+        # rubocop:enable Lint/UnneededCopDisableDirective
+        MSG = 'Comment to disable/enable Rubocop.'
```
[the remark this answers is not in the document]
Comment (egze) [reply lib/rubocop/cop/style/disable_cops_within_source_code_directive.rb]: Done

**Comment by egze** [comment]
> So, I like the cop and the changes look good (had some minor comments), and I would like to see the `DisabledLinesFormatter` gone as well.

Removed `DisabledLinesFormatter` as well.

**Comment by bbatsov** [comment]
> I love the idea! In fact, I argued rather stubbornly for implementing this kind of functionality in a cop back in #900, but I was overruled and and the --format disabled CLI option was added instead. This option doesn't currently work (I just discovered), and hasn't worked for a long time. That means that people aren't using it.

Amazing memory! I had totally forgotten about this!

> So, I like the cop and the changes look good (had some minor comments), and I would like to see the DisabledLinesFormatter gone as well.

Agreed!

File: CHANGELOG.md [hunk 3]
```diff
@@ -197,6 +197,7 @@
 
 ### New features
 
+* [#7384](https://github.com/rubocop-hq/rubocop/pull/7384): Add new `Style/DisableCopsWithinSourceCodeDirective` cop. ([@egze][])
```
Comment (jonas054) [inline CHANGELOG.md]: This line needs to be moved up to the `master (unreleased)` section.

Comment (egze) [reply CHANGELOG.md]: Done

Comment (jonas054) [reply CHANGELOG.md]: Sorry, I should have been more clear. It should actually go under "New features" and you can put it last in the list of (unreleased) new features.

Comment (egze) [reply CHANGELOG.md]: Got it :) done

**Pull request merged**